server/cluster/cluster.go

```diff
@@ -665,6 +664,58 @@ func (c *RaftCluster) processRegionHeartbeat(region *core.RegionInfo) error {
 		time.Sleep(500 * time.Millisecond)
 	})
 
+	return c.saveRegionLock(isNew, saveKV, saveCache, needSync, region, origin)
+}
+
+// processRegionHeartbeatResponse updates the region information thought region split.
+// It will update peer that not exist in the cache, and add new peers that join in.
+func (c *RaftCluster) processRegionSpiltReport(regions []*metapb.Region) (err error) {
+	if err = checkSplitRegions(regions); err != nil {
+		return err
+	}
+	c.RLock()
+	coreCluster := c.core
+	c.RUnlock()
+	total := len(regions) - 1
+	for i := range regions {
+		// It should update rightmost because it will override by left .
+		v := regions[total-i]
+		if v.Peers == nil {
+			err = errors.Wrap(err, "region peer is nil")
+			continue
+		}
+		region := core.NewRegionInfo(v, v.Peers[0])
+		origin, perr := coreCluster.PreCheckPutRegion(region)
+		if perr != nil {
+			err = errors.Wrap(err, perr.Error())
+			continue
+		}
+		update, isNew, needSync := false, false, true
+		// The region that is not the rightmost element should update if origin is nil.
+		// The rightmost region should update if epoch greater than origin.
+		if origin == nil {
+			update = true
+		} else if i == len(regions)-1 {
+			isNew = false
+			r, o := region.GetRegionEpoch(), origin.GetRegionEpoch()
+			if r.GetVersion() > o.GetVersion() || r.GetConfVer() > o.GetConfVer() {
+				update = true
+				needSync = false
+			}
+		}
+		if update {
+			if e := c.saveRegionLock(isNew, true, true, needSync, region, origin); e != nil {
+				err = errors.Wrap(err, e.Error())
+			}
+		}
+	}
+	return err
+}
+
+func (c *RaftCluster) saveRegionLock(isNew, saveKV, saveCache, needSync bool, region *core.RegionInfo, origin *core.RegionInfo) error {
```

Comment:
The name `xxxLock` or `xxxLocked` should only be used when it's in the situation that the caller handles the lock.

Reply:
yes, it will rename as saveRegion